File: src/test/java/com/runninglane/dto/buddy/test/java/cases/abstracts/SimpleAbstractDtoTest.java
package com.runninglane.dto.buddy.test.java.cases.abstracts;

import com.runninglane.dto.buddy.javainterop.DtoBuddy;
import com.runninglane.dto.buddy.test.NamingStrategyWithCountUpSuffix;
import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class SimpleAbstractDtoTest {
    private final DtoBuddy dtoBuddy = new DtoBuddy(new NamingStrategyWithCountUpSuffix());

    @Test
    public void testImplement() {
        Class<?> concreteClass = dtoBuddy.implement(SimpleAbstractDto.class);

        assertNotNull(concreteClass);
        assertTrue(SimpleAbstractDto.class.isAssignableFrom(concreteClass));
    }

    @Test
    public void testCreate() {
        // This test will fail until create() is implemented
        Map<String, Object> params = new HashMap<>();
        params.put("name", "John Doe");
        params.put("age", 30);
        params.put("email", "john.doe@example.com");

        Class<?> concreteClass = dtoBuddy.implement(SimpleAbstractDto.class);
        SimpleAbstractDto dto = dtoBuddy.create(concreteClass, params);

        assertNotNull(dto);
        assertEquals("John Doe", dto.getName());
        assertEquals(30, dto.getAge());
        assertEquals("john.doe@example.com", dto.getEmail());
    }

    @Test
    public void testCreateThenPopulate() {
        // This test will fail until populate() is implemented
        Map<String, Object> params = new HashMap<>();
        params.put("email", "john.doe@example.com");

        Class<?> concreteClass = dtoBuddy.implement(SimpleAbstractDto.class);
        Map<String, Object> initialParams = new HashMap<>();
        initialParams.put("name", "John Doe");
        initialParams.put("age", 30);
        initialParams.put("email", null);
        SimpleAbstractDto dto = dtoBuddy.create(concreteClass, initialParams);

        assertNotNull(dto);
        dtoBuddy.populate(dto, params);
        assertEquals("John Doe", dto.getName());
        assertEquals(30, dto.getAge());
        assertEquals("john.doe@example.com", dto.getEmail());
    }

    @Test
    public void testCreateWithNullValues() {
        // This test will fail until create() is implemented
        Map<String, Object> params = new HashMap<>();
        params.put("name", "Jane Doe");
        params.put("age", 25);
        params.put("email", null);

        Class<?> concreteClass = dtoBuddy.implement(SimpleAbstractDto.class);
        SimpleAbstractDto dto = dtoBuddy.create(concreteClass, params);

        assertNotNull(dto);
        assertEquals("Jane Doe", dto.getName());
        assertEquals(25, dto.getAge());
        assertNull(dto.getEmail());
    }

    @Test
    public void testCreateThenPopulateWithMultipleValues() {
        // This test will fail until populate() is implemented
        Map<String, Object> params = new HashMap<>();
        params.put("age", 30);
        params.put("email", "john.doe@example.com");

        Class<?> concreteClass = dtoBuddy.implement(SimpleAbstractDto.class);
        Map<String, Object> initialParams = new HashMap<>();
        initialParams.put("name", "John Doe");
        initialParams.put("age", 10);
        initialParams.put("email", null);
        SimpleAbstractDto dto = dtoBuddy.create(concreteClass, initialParams);

        assertNotNull(dto);

        dtoBuddy.populate(dto, params);

        assertEquals("John Doe", dto.getName());
        assertEquals(30, dto.getAge());
        assertEquals("john.doe@example.com", dto.getEmail());
    }
}

File: src/test/kotlin/com/runninglane/dto/buddy/test/cases/abstracts/SimpleAbstractDtoTest.kt
package com.runninglane.dto.buddy.test.cases.abstracts

import com.runninglane.dto.buddy.DtoBuddy
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class SimpleAbstractDtoTest {
    companion object {
        private var nameSuffix: Int = 0
    }

    @Test
    fun testImplement() {
         val concreteClass = DtoBuddy.implement(SimpleAbstractDto::class.java)

         assertNotNull(concreteClass)
         assertTrue(SimpleAbstractDto::class.java.isAssignableFrom(concreteClass))
    }

    @Test
    fun testCreate() {
        // This test will fail until create() is implemented
        val params = mapOf(
            "name" to "John Doe",
            "age" to 30,
            "email" to "john.doe@example.com"
        )

        val concreteClass = DtoBuddy.implement(SimpleAbstractDto::class.java, nameSuffix = "${++nameSuffix}")
        val dto: SimpleAbstractDto = DtoBuddy.create(concreteClass, params)

        assertNotNull(dto)
        assertEquals("John Doe", dto.name)
        assertEquals(30, dto.age)
        assertEquals("john.doe@example.com", dto.email)
    }

    @Test
    fun testCreateThenPopulate() {
        // This test will fail until populate() is implemented
        val params = mapOf(
            "email" to "john.doe@example.com"
        )

        val concreteClass = DtoBuddy.implement(SimpleAbstractDto::class.java, nameSuffix = "${++nameSuffix}")
        val dto: SimpleAbstractDto = DtoBuddy.create(
            concreteClass,
            mapOf("name" to "John Doe", "age" to 30, "email" to null)
        )

        assertNotNull(dto)
        DtoBuddy.populate(dto, params)
        assertEquals("John Doe", dto.name)
        assertEquals(30, dto.age)
        assertEquals("john.doe@example.com", dto.email)
    }

    @Test
    fun testCreateWithNullValues() {
        // This test will fail until create() is implemented
        val params = mapOf(
            "name" to "Jane Doe",
            "age" to 25,
            "email" to null
        )

        val concreteClass = DtoBuddy.implement(SimpleAbstractDto::class.java, nameSuffix = "${++nameSuffix}")
        val dto: SimpleAbstractDto = DtoBuddy.create(concreteClass, params)

        assertNotNull(dto)
        assertEquals("Jane Doe", dto.name)
        assertEquals(25, dto.age)
        assertEquals(null, dto.email)
    }

    @Test
    fun testCreateThenPopulateWithMultipleValues() {
        // This test will fail until populate() is implemented
        val params = mapOf(
            "age" to 30,
            "email" to "john.doe@example.com"
        )

        val concreteClass = DtoBuddy.implement(SimpleAbstractDto::class.java, nameSuffix = "${++nameSuffix}")
        val dto: SimpleAbstractDto = DtoBuddy.create(
            concreteClass,
            mapOf("name" to "John Doe", "age" to 10, "email" to null)
        )

        assertNotNull(dto)

        DtoBuddy.populate(dto, params)

        assertEquals("John Doe", dto.name)
        assertEquals(30, dto.age)
        assertEquals("john.doe@example.com", dto.email)
    }
}